### PR TITLE
two-factor activities

### PR DESCRIPTION
**Pull request by ChristophWurst**
This adds generic support for two-factor activities (setting, filter and provider). The backup codes provider is the first that sends custom events when the user creates codes. I will add enabled/disable events to TOTP and U2F too. I added the generic classes to the backup provider too because those are registered per app and it's shipped with the server.

* Publishes an activity when a 2fa attempt failed
* Publishes an activity when a 2fa attempf succeeded
* Publishes an activity when backup codes are generated

@LukasReschke as requested :-)

cc @nickvergessen 

## Screenshot
bildschirmfoto von 2016-12-13 10-48-42

### CODE REVIEW

File: apps/twofactor_backupcodes/lib/Activity/GenericFilter.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	/** @var IL10N */
+	private $l10n;
+
+	public function __construct(IURLGenerator $urlGenerator, IL10N $l10n) {
+		$this->urlGenerator = $urlGenerator;
+		$this->l10n = $l10n;
+	}
+
+	public function allowedApps() {
+		return null;
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Activity/GenericFilter.php]: No, **must** return array. Use an empty one when you want to allow any app

File: apps/twofactor_backupcodes/lib/Activity/Provider.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	public function parse($language, IEvent $event, IEvent $previousEvent = null) {
+		if ($event->getApp() !== 'twofactor_backupcodes') {
+			throw new InvalidArgumentException();
+		}
+
+		$l = $this->l10n->get('twofactor_backupcodes', $language);
+
+		switch ($event->getSubject()) {
+			case 'codes_generated':
+				$event->setParsedSubject($l->t('You created backup codes for your account'));
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Activity/Provider.php]: what backup codes? I guess we should mention 2FA in this sentence

File: apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php [hunk 3]
```diff
@@ -68,10 +73,28 @@ public function createCodes(IUser $user, $number = 10) {
...
+	 * @param IUser $user
+	 * @param string $event
+	 */
+	private function publishEvent(IUser $user, $event) {
+		$activity = $this->activityManager->generateEvent();
+		$activity->setApp('twofactor_backupcodes')
+			->setType('twofactor')
+			->setAuthor($user->getUID())
+			->setAffectedUser($user->getUID());
+		$activity->setSubject($event);
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: chain as well?

File: apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php [hunk 4]
```diff
@@ -68,10 +73,28 @@ public function createCodes(IUser $user, $number = 10) {
...
+		$this->activityManager->publish($activity);
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: Please catch potential exception and log it, so it doesn't break the authentication in case something ever goes wrong

File: lib/private/Authentication/TwoFactorAuth/Manager.php [hunk 5]
```diff
@@ -184,11 +190,35 @@ public function verifyChallenge($providerId, IUser $user, $challenge) {
...
+	 * @param string $event
+	 */
+	private function publishEvent(IUser $user, $event, array $params) {
+		$activity = $this->activityManager->generateEvent();
+		$activity->setApp('twofactor_generic')
+			->setType('twofactor_generic')
+			->setAuthor($user->getUID())
+			->setAffectedUser($user->getUID());
+		$activity->setSubject($event, $params);
+		$this->activityManager->publish($activity);
```
Comment (nickvergessen) [inline lib/private/Authentication/TwoFactorAuth/Manager.php]: same here, chain + catch

File: lib/private/Authentication/TwoFactorAuth/Manager.php [hunk 6]
```diff
@@ -184,11 +190,35 @@ public function verifyChallenge($providerId, IUser $user, $challenge) {
...
 	/**
+	 * Push a 2fa event the user's activity stream
+	 *
+	 * @param IUser $user
+	 * @param string $event
+	 */
+	private function publishEvent(IUser $user, $event, array $params) {
+		$activity = $this->activityManager->generateEvent();
+		$activity->setApp('twofactor_generic')
+			->setType('twofactor_generic')
```
Comment (nickvergessen) [inline lib/private/Authentication/TwoFactorAuth/Manager.php]: Provider checks `if ($event->getType() !== 'twofactor') {`

Comment (nickvergessen) [inline lib/private/Authentication/TwoFactorAuth/Manager.php]: Also `twofactor` is the registered setting, not `*_generic`

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php (hunk 4 above)
Comment (ChristophWurst) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: which one? The API doc does not mention any specific exception. Catching \Exception is bad IMO because it potentially swallows unrelated stuff and bugs go unnoticed :-/

**Comment by ChristophWurst** [comment]
@nickvergessen thanks for your review. I've pushed another commit to fix the issues you noticed in the review comments.

File: apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php [hunk 7]
```diff
@@ -89,9 +96,14 @@ private function publishEvent(IUser $user, $event) {
...
 			->setType('twofactor')
 			->setAuthor($user->getUID())
-			->setAffectedUser($user->getUID());
-		$activity->setSubject($event);
-		$this->activityManager->publish($activity);
+			->setAffectedUser($user->getUID())
+			->setSubject($event);
+		try {
+			$this->activityManager->publish($activity);
+		} catch (Exception $e) {
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: \BadMethodCallException

Comment (ChristophWurst) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: why?

Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: Because that is what is thrown, no need to catch everything 😉

Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: I just noticed that the `@throws` is missing from the public interface.

Comment (ChristophWurst) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: should we add it?

**Comment by ChristophWurst** [comment]
@nickvergessen please re-review :-)

File: apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php (hunk 7 above)
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: Yes please

Comment (ChristophWurst) [inline apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php]: done https://github.com/nextcloud/server/pull/2754

File: apps/twofactor_backupcodes/lib/Activity/GenericProvider.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+			case 'twofactor_success':
+				$params = $event->getSubjectParameters();
+				$event->setParsedSubject($l->t('You successfully logged in using two-factor authentication (%1$s)', [
+							$params['provider'],
+					]));
+				$event->setIcon($this->urlGenerator->getAbsoluteURL($this->urlGenerator->imagePath('core', 'actions/password.svg')));
+				break;
+			case 'twofactor_failed':
+				$params = $event->getSubjectParameters();
+				$event->setParsedSubject($l->t('A login attempt using two-factor authenticatoin failed (%1$s)', [
```
Comment (nickvergessen) [inline apps/twofactor_backupcodes/lib/Activity/GenericProvider.php]: typo in `authenticatooooooooooiiiiiiiiiiiiiiiiiiiiiin`

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Looks good apart from the type
👍

**Review by nickvergessen — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
I retriggered CI.
